charger/bender.go

```diff
@@ -0,0 +1,261 @@
+
+const (
+	// all holding type registers
+	bndRegFirmware         = 100  // Application version number
+	bndRegChargePointState = 122  // Vehicle (Control Pilot) state
+	bndRegChargePointModel = 142  // ChargePoint Model. Bytes 0 to 19.
+	bndRegCurrentL1        = 212  // Current (phase 1) from primary meter (mA)
+	bndRegCurrentL2        = 214  // Current (phase 2) from primary meter (mA)
+	bndRegCurrentL3        = 216  // Current (phase 3) from primary meter (mA)
+	bndRegTotalEnergy      = 218  // Total Energy from primary meter (Wh)
+	bndRegActivePower      = 220  // Active Power from primary meter (W)
+	bndRegChargedEnergy    = 716  // Sum of charged energy for the current session (Wh)
+	bndRegChargingDuration = 718  // Duration since beginning of charge (Seconds)
+	bndRegEVCCID           = 741  // ASCII representation of the Hex. Values corresponding to the EVCCID. Bytes 0 to 11.
+	bndRegHemsCurrentLimit = 1000 // Current limit of the HEMS module (A)
+)
+
+var bndRegCurrents = []uint16{bndRegCurrentL1, bndRegCurrentL2, bndRegCurrentL3}
+
+func init() {
+	registry.Add("bender", NewBenderCCFromConfig)
+}
+
+// NewBenderCCFromConfig creates a BenderCC charger from generic config
+func NewBenderCCFromConfig(other map[string]interface{}) (api.Charger, error) {
+	cc := modbus.TcpSettings{
+		ID: 255,
+	}
+
+	if err := util.DecodeOther(other, &cc); err != nil {
+		return nil, err
+	}
+
+	return NewBenderCC(cc.URI, cc.ID)
+}
+
+// NewBenderCC creates BenderCC charger
+func NewBenderCC(uri string, id uint8) (api.Charger, error) {
+	conn, err := modbus.NewConnection(uri, "", "", 0, modbus.Tcp, id)
+	if err != nil {
+		return nil, err
+	}
+
+	if !sponsor.IsAuthorized() {
+		return nil, api.ErrSponsorRequired
+	}
+
+	log := util.NewLogger("bender")
+	conn.Logger(log.TRACE)
+
+	wb := &BenderCC{
+		conn:    conn,
+		current: 6, // assume min current
+	}
+
+	return wb, err
+}
+
+// Status implements the api.Charger interface
+func (wb *BenderCC) Status() (api.ChargeStatus, error) {
+	b, err := wb.conn.ReadHoldingRegisters(bndRegChargePointState, 1)
+	if err != nil {
+		return api.StatusNone, err
+	}
+
+	sb := binary.BigEndian.Uint16(b)
+
+	switch sb {
+	case 1:
+		return api.StatusA, nil
+	case 2:
+		return api.StatusB, nil
+	case 3:
+		return api.StatusC, nil
+	case 4:
+		return api.StatusD, nil
+	case 5:
+		return api.StatusE, nil
+	default:
+		return api.StatusNone, fmt.Errorf("invalid status: %d", sb)
+	}
+}
+
+// Enabled implements the api.Charger interface
+func (wb *BenderCC) Enabled() (bool, error) {
+	b, err := wb.conn.ReadHoldingRegisters(bndRegHemsCurrentLimit, 1)
+	if err != nil {
+		return false, err
+	}
+
+	cur := binary.BigEndian.Uint16(b)
+
+	enabled := cur != 0
+	if enabled {
+		wb.current = cur
+	}
+
+	return enabled, nil
+}
+
+// Enable implements the api.Charger interface
+func (wb *BenderCC) Enable(enable bool) error {
+	var cur uint16
+	if enable {
+		cur = wb.current
+	}
+
+	b := make([]byte, 2)
+	binary.BigEndian.PutUint16(b, cur)
```

Comment:
Das kannst Du auch in den if enable Zweig verschieben. Das slice ist immer mit Null initialisiert.